Update Map Guide (#6345)
sunpy/sunpy: docs/guide/data_types/maps.rst

This class offers many advantages over using, for example, `astropy.io.fits` to open 2D solar data in Python.
SunPy Maps recognize data from 19 sources (see :doc:`/code_ref/map` for a complete list) and can provide instrument-specific information when plotting and even correct/add metadata when it is missing from the original file.

Part of the philosophy of the Map object is to provide most of the basic functionality that a scientist would want.
Therefore, a Map also contains a number of methods such as resizing or grabbing a subview.
See `~sunpy.map.mapbase.GenericMap` for summaries of all attributes and methods.

:ref:`creating-maps` and :ref:`inspecting-maps` demonstrate how to create a Map object and view a summary of the data in the object.
:ref:`map-data` describes how data is stored and accessed in a Map, and :ref:`plotting-maps` introduces some basics of visualizing Map data in SunPy.
:ref:`composite-maps` and :ref:`map-sequences` detail how SunPy can handle multiple Maps from similar and different instruments.
Lastly, :ref:`custom-maps` shows how you can create Maps from data sources not supported in SunPy.

.. _creating-maps:

> 1. Creating Maps

Reviewer: We in theory can have sphinx auto number it for us. https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html#ordered-lists

Thoughts?

*********
Map Guide
*********

Map objects in **sunpy** are two-dimensional data associated with a coordinate system.
This class offers many advantages over using packages such as `astropy.io.fits` to open 2D solar data in Python.

> **sunpy** Maps recognize data from 19 sources (see :doc:`/code_ref/map` for a complete list) and can provide instrument-specific information when plotting and even correct or add metadata when it is missing from the original file.

Reviewer 2: ```suggestion
**sunpy** Maps can be used with any image with two spatial axes that has FITS standard compliant metadata.
There are also 19 sources (see :doc:`/code_ref/map` for a complete list) that provide instrument-specific information when plotting and in some cases interpret missing metadata when it is missing from the original file.
```
I we should be clear that it's not only maps with map sources that can be opened? Also we deliberately don't add any missing metadta, just guess it 😃

> **sunpy** Maps recognize data from 19 sources (see :doc:`/code_ref/map` for a complete list) and can provide instrument-specific information when plotting and even correct or add metadata when it is missing from the original file.

Reviewer 2: This is a superb introduction!

          ('cunit1', Unit("arcsec")),
          ('cunit2', Unit("arcsec")),
          ('ctype1', 'HPLN-TAN'),
          ('ctype2', 'HPLT-TAN'),
          ('crval1', 0.0),
          ('crval2', 0.0),
          ...
          ('date-obs', '2013-10-28T00:00:00.000'),
          ('hgln_obs', 0.0),
          ('hglt_obs', 4.7711570596394015),
          ('dsun_obs', 148644585949.4918),
          ('rsun_ref', 695700.0),
          ('rsun_obs', 965.3829548285768),
          ('instrume', 'Test case'),
          ('wavelnth', 1000),
          ('detector', 'UV detector'),
          ('waveunit', 'angstrom')])

From these header MetaDict's that are generated, we can now create a custom map::

    >>> my_map = sunpy.map.Map(data, header) # doctest: +SKIP
    >>> my_map.peek() # doctest: +SKIP
2. Inspecting Maps
==================

Inspecting maps
===============
A map contains a number of data-associated attributes. To get a quick look at
your map simply type::
A Map contains a number of data-associated attributes. To get a quick look at
your Map simply type: ::

    >>> my_map = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)  # doctest: +REMOTE_DATA
    >>> my_map  # doctest: +REMOTE_DATA

>     <sunpy.map.sources.sdo.AIAMap object at ...>

Reviewer: Was there a reason to remove the output from this?

Author: My thinking is that a person using the guide will have the guide open in one window and a python console in another (where they will be typing and executing each step of the guide). So, they will execute the command and read the paragraph following this code which explains what they are looking at. In the case of plots, I think it is helpful to have an example a user can reference but the summary output just seemed to take up space.

Reviewer: So the issue is that these code blocks are executed as tests, so you will need to also remove the line above to stop it expecting that output to be there.

Author: Okay, I think I understand. Could I remove line 65 and change line 66 to: `>>> my_map  # doctest: +SKIP`?

Reviewer: I am not in favour of skipping lines unless we have a good reason. I would rather remove them.

Personally, I don't think we should be removing the output from this but I will want others to chime in first.

Reviewer 3: I think it's a different take on what a "guide" is.

Author: After giving this some thought, I think the output should be included. While I do approach this guide differently, it should stay consistent with the rest of the documentation.

           [-128.03072  , -128.03072  , -128.03072  , ..., -128.03072  ,
            -128.03072  , -128.03072  ]], dtype=float32)

This will show a representation of the data as well as some of its associated attributes.
Typing the above command in a Jupyter Notebook will show a rich HTML view of the table along with two plots of your data.

> The HTML view can also be accessed using the :func:`~sunpy.map.GenericMap.quicklook` function, which will open the view in your default browser.

Reviewer 2: ```suggestion
The HTML view can also be accessed using the :func:`~sunpy.map.GenericMap.quicklook` method, which will open the view in your default browser.
```

The full list can be found on `~sunpy.map.GenericMap`:

.. code-block:: python

    >>> map_date = my_map.date  # doctest: +REMOTE_DATA
    >>> map_exptime = my_map.exposure_time  # doctest: +REMOTE_DATA
    >>> map_center = my_map.center  # doctest: +REMOTE_DATA

To get a list of all of the attributes check the documentation by typing:

.. code-block:: python

    >>> help(my_map)  # doctest: +SKIP

Many attributes and functions of the map classes accept and return `~astropy.units.quantity.Quantity` or `~astropy.coordinates.SkyCoord` objects.
Please refer to :ref:`units-coordinates-sunpy` for more details.

The metadata for the map is accessed by:

.. code-block:: python

    >>> header = my_map.meta  # doctest: +REMOTE_DATA

This references the metadata dictionary with the header information as read from the source file.
To see if the metadata of a Map source has been modified, see :ref:`sphx_glr_generated_gallery_map_map_metadata_modification.py` for a demonstration.

.. _map-data:

3. Map Data
===========

The data in a Map object is accessible through the `~sunpy.map.GenericMap.data` attribute.

> The data is implemented as a NumPy `~numpy.ndarray`.

Reviewer 2: ```suggestion
The data is stored as a NumPy `~numpy.ndarray`.
```

For example, to get the 0th element in the array:

.. code-block:: python

    >>> my_map.data[0, 0]  # doctest: +REMOTE_DATA
    -95.92475
    >>> my_map.data[0][0]  # doctest: +REMOTE_DATA
    -95.92475

The first index is for the y direction while the second index is for the x direction.
For more information about indexing, please refer to the `numpy documentation <https://numpy.org/doc/stable/user/basics.indexing.html#indexing-on-ndarrays>`__.

Data attributes like `~numpy.ndarray.dtype` and `~sunpy.map.GenericMap.dimensions` are accessible through a GenericMap object:

.. code-block:: python

    >>> my_map.dimensions  # doctest: +REMOTE_DATA
    PixelPair(x=<Quantity 1024. pix>, y=<Quantity 1024. pix>)
    >>> my_map.dtype  # doctest: +REMOTE_DATA
    dtype('float32')

Here, the dimensions attribute is similar to the `~numpy.ndarray.shape` attribute, however returning an `~astropy.units.quantity.Quantity`.

You can store the data of a `~sunpy.map.GenericMap` object in a separate `~numpy.ndarray` by either of the following actions:

.. code-block:: python

    >>> var = my_map.data  # doctest: +REMOTE_DATA
    >>> var = my_map.data.copy()  # doctest: +REMOTE_DATA

> To create a complete copy of a Map object that is entirely independent of the original, use the built-in `copy.deepcopy` method:

Reviewer 2: ```suggestion
To create a complete copy of a Map object that is entirely independent of the original, use the built-in `copy.deepcopy` function:
```

    >>> my_map.mean()  # doctest: +REMOTE_DATA
    427.02252

All the other `~numpy.ndarray` functions and attributes can be accessed through the data array directly.
For example:

.. code-block:: python

    >>> my_map.data.std()  # doctest: +REMOTE_DATA
    826.41016

.. _plotting-maps:

4. Plotting Maps
================

The `~sunpy.map.GenericMap` object has a built-in plot method such that it is easy to quickly view your map.
To create a plot just type:

.. code-block:: python

    >>> my_map.peek()   # doctest: +SKIP

> This will open a matplotlib plot on your screen.

Reviewer 2: ```suggestion
This will open a Matplotlib plot on your screen.
```

> In addition, it is possible to grab the matplotlib axes object by using the `~sunpy.map.GenericMap.plot()` command.

Reviewer 2: ```suggestion
In addition, it is possible to grab the Matplotlib Axes object by using the `~sunpy.map.GenericMap.plot()` command.
```

    import sunpy.data.sample
    import matplotlib.pyplot as plt
    smap = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)
    fig = plt.figure()
    smap.plot(cmap=plt.cm.Greys_r)
    plt.colorbar()
    plt.show()

You can view or make changes to the default settings through the ``sunpy.map.GenericMap.plot_settings`` dictionary.
See `Editing the colormap and normalization of a Map <https://docs.sunpy.org/en/stable/generated/gallery/plotting/map_editcolormap.html>`_ for an example of this method.

4.2 Colormaps and Normalization
-------------------------------

Image data is generally shown in false color in order to better identify it or to better visualize structures in the image.
Matplotlib handles this colormapping process through the `~matplotlib.colors` module.
First, the data array is mapped onto the range 0-1 using an instance of `~matplotlib.colors.Normalize` or a subclass.

> Then, the data is mapped to a color using a `~matplotlib.colors.Colormap`.

Author: I agree. It was for this reason that I linked to five "foundational" examples at the beginning of the Plotting Maps section. Since those examples cover many of the basics for plotting in short self-contained chunks, I see these plotting subsections as additional information for more advanced users.

The Map object chooses the appropriate colormap for you when it is created as
long as it recognizes the instrument. To see what colormaps are available::
SunPy provides colormaps for each mission as defined by the mission teams.
The Map object chooses the appropriate colormap for you when it is created as long as it recognizes the instrument.
To see what colormaps are available: ::

    >>> import sunpy.visualization.colormaps as cm
    >>> cm.cmlist.keys()
    dict_keys(['goes-rsuvi94', 'goes-rsuvi131', 'goes-rsuvi171', 'goes-rsuvi195',
    'goes-rsuvi284', 'goes-rsuvi304', 'sdoaia94', 'sdoaia131', 'sdoaia171',
    ...

The sunpy colormaps are registered with matplotlib so you can grab them like
you would any other colormap::
The SunPy colormaps are registered with matplotlib so you can grab them like you would any other colormap: ::

    >>> import matplotlib.pyplot as plt
    >>> import sunpy.visualization.colormaps

You need to import `sunpy.visualization.colormaps` or `sunpy.map` for this to work::

    >>> cmap = plt.get_cmap('sdoaia171')

See `~sunpy.visualization.colormaps` for a plot of all available colormaps.

> The following plot shows off all of the colormaps.

Author: The plot of all the colormaps is linked to via `~sunpy.visualization.colormaps`.

> The following plot shows off all of the colormaps.

Reviewer: I missed that!

    import sunpy.data.sample
    import matplotlib.pyplot as plt
    smap = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)
    fig = plt.figure()
    smap.plot(cmap=plt.cm.Greys_r)
    plt.colorbar()
    plt.show()

> You can view or make changes to the default settings through the ``sunpy.map.GenericMap.plot_settings`` dictionary.

Reviewer 2: ```suggestion
You can also view or make changes to the default settings through the ``sunpy.map.GenericMap.plot_settings`` dictionary.
```

> See :ref:`sphx_glr_generated_gallery_plotting_map_editcolormap.py` for an example of this method.

Reviewer 2: ```suggestion
See :ref:`sphx_glr_generated_gallery_plotting_map_editcolormap.py` for an example of this workflow for changing plot settings.
```

4.2 Colormaps and Normalization
-------------------------------

Image data is generally shown in false color in order to better identify it or to better visualize structures in the image.
Matplotlib handles this colormapping process through the `~matplotlib.colors` module.
First, the data array is mapped onto the range 0-1 using an instance of `~matplotlib.colors.Normalize` or a subclass.
Then, the data is mapped to a color using a `~matplotlib.colors.Colormap`.

**sunpy** provides colormaps for each mission as defined by the mission teams.
The Map object chooses the appropriate colormap for you when it is created as long as it recognizes the instrument.
To see what colormaps are available:

.. code-block:: python

    >>> import sunpy.visualization.colormaps as cm
    >>> cm.cmlist.keys()
    dict_keys(['goes-rsuvi94', 'goes-rsuvi131', 'goes-rsuvi171', 'goes-rsuvi195',
    'goes-rsuvi284', 'goes-rsuvi304', 'sdoaia94', 'sdoaia131', 'sdoaia171',
    ...

> The **sunpy** colormaps are registered with matplotlib so you can grab them like you would any other colormap:

Reviewer 2: ```suggestion
The **sunpy** colormaps are registered with Matplotlib so you can grab them like you would any other colormap:
```

    plt.colorbar()
    plt.show()

You can also change the colormap for the Map itself:

.. code-block:: python

    >>> smap.plot_settings['cmap'] = plt.get_cmap('sohoeit171')  # doctest: +SKIP

The normalization is set automatically so that all the data from minimum to maximum is displayed as best as possible.
Just like the colormap, the default normalization can be changed through the ``plot_settings`` dictionary or directly for the individual plot by passing a keyword argument.

>                

Reviewer: ```suggestion

```

* `Overplotting HMI Contours on an AIA Image <https://docs.sunpy.org/en/stable/generated/gallery/map/hmi_contours_wcsaxes.html>`_

For a more advanced tutorial on combining data from several maps, see `Creating a Full Sun Map with AIA and EUVI <https://docs.sunpy.org/en/stable/generated/gallery/map_transformations/reprojection_aia_euvi_mosaic.html>`_.

.. _map-sequences:

6. Map Sequences
================

A `~sunpy.map.MapSequence` is an ordered list of maps.
By default, the maps are ordered by their observation date, from earliest to latest date.
A `~sunpy.map.MapSequence` can be created by supplying multiple existing maps: ::

    >>> map1 = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)  # doctest: +REMOTE_DATA
    >>> map2 = sunpy.map.Map(sunpy.data.sample.EIT_195_IMAGE)  # doctest: +REMOTE_DATA
    >>> mc = sunpy.map.Map([map1, map2], sequence=True)  # doctest: +REMOTE_DATA

or by providing a directory full of image files: ::

    >>> mc = sunpy.map.Map('path/to/my/files/*.fits', sequence=True)   #  doctest: +SKIP

The earliest map in the MapSequence can be accessed by indexing the maps list: ::

>     >>> mc.maps[0]   # doctest: +SKIP

Author: I think it was skipped because the output from this command is identical to the map summary output included in the Inspecting Maps section.

MapSequences can hold maps that have different shapes.
To test if all the maps in a `~sunpy.map.MapSequence` have the same shape: ::

    >>> mc.all_maps_same_shape()  # doctest: +REMOTE_DATA
    True

It is often useful to return the image data in a `~sunpy.map.MapSequence` as a single three dimensional NumPy `~numpy.ndarray`: ::

    >>> mc.as_array()   # doctest: +SKIP

Note that an array is returned only if all the maps have the same shape.
If this is not true, a ``ValueError`` is returned.
If all the maps have nx pixels in the x-direction, and ny pixels in the y-direction, and there are n maps in the MapSequence, the returned `~numpy.ndarray` array has shape (ny, nx, n).
The data of the first map in the `~sunpy.map.MapSequence` appears in the `~numpy.ndarray` in position ``[:, :, 0]``, the data of second map in position ``[:, :, 1]``, and so on.
The order of maps in the `~sunpy.map.MapSequence` is reproduced in the returned `~numpy.ndarray`.

The metadata from each map can be obtained using: ::

    >>> mc.all_meta()   # doctest: +SKIP

This returns a list of map meta objects that have the same order as the maps in the `~sunpy.map.MapSequence`.

> 6.1 Coalignment of Map Sequences

Author: As an idea, the coalignment and solar rotation map sequence sections could be adapted into examples for the sunkit_image example gallery, and I can leave a note here to point users to the `sunkit_image.coalignment` module.

> 6.1 Coalignment of Map Sequences

Reviewer: This makes a ton of sense. Link to that example gallery here and I can make examples from these.

> 6.1 Coalignment of Map Sequences

Author: I just pushed a large amount of changes to this PR. Unless you have already started working on these examples, I'd be happy to write them in the meantime.

> 6.1 Coalignment of Map Sequences

Reviewer: If you would, that would be great.

    pc2_1: 0.0
    pc2_2: 1.0
    rsun_obs: 965.3829548285768

As we can see, a list of WCS and observer meta information is contained within the generated headers, however we may want to include other meta information including the observatory name, the wavelength and waveunit of the observation.
Any of the keywords listed in ``header_helper.meta_keywords`` can be passed to the `~sunpy.map.make_fitswcs_header` and will then populate the returned MetaDict header.
Furthermore, the following observation keywords can be passed to the `~sunpy.map.make_fitswcs_header` function and will be translated to the FITS standard: ``observtory``, ``instrument``, ``telescope``, ``wavelength``, ``exposure``.

An example of creating a header with these additional keywords: ::

    >>> header = sunpy.map.make_fitswcs_header(data, coord,
    ...                                        reference_pixel = u.Quantity([5, 5]*u.pixel),
    ...                                        scale = u.Quantity([2, 2] *u.arcsec/u.pixel),
    ...                                        telescope = 'Test case', instrument = 'UV detector',
    ...                                        wavelength = 1000*u.angstrom)

>     >>> header  # doctest: +SKIP

Author: This line is currently skipped in the guide. I am not sure why it is skipped since the output is there.

    >>> coaligned = mapsequence_coalign_by_match_template(mc)  # doctest: +REMOTE_DATA,+IGNORE_WARNINGS

This will return a new `~sunpy.map.MapSequence`, coaligned to a template extracted from the
center of the first map in the `~sunpy.map.MapSequence`, with the map dimensions clipped as
required.  The coalignment algorithm provides many more options for handling
the coalignment of `~sunpy.map.MapSequence`. Type: ::

    >>> help(mapsequence_coalign_by_match_template)   # doctest: +SKIP

for a full list of options and functionality.

If you want to calculate the shifts required to compensate for solar
rotation relative to the first map in the `~sunpy.map.MapSequence` without applying them, use: ::

    >>> from sunpy.image.coalignment import calculate_match_template_shift
    >>> shifts = calculate_match_template_shift(mc)  # doctest: +REMOTE_DATA,+IGNORE_WARNINGS

This is the algorithm used by the coalignment function to calculate the shifts.
See `~sunpy.image.coalignment.calculate_match_template_shift` to learn more about its features.
Shifts calculated using calculate_match_template_shift can be passed directly
to the coalignment function.

> For similar examples, see the `Combing, Co-aligning, and Reprojecting Images <https://docs.sunpy.org/en/stable/generated/gallery/index.html#combining-co-aligning-and-reprojecting-images>`_

Reviewer 3: "Combing" -> "Combining"

A `~sunpy.map.MapSequence` can be created by supplying multiple existing maps:

.. code-block:: python

    >>> map1 = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)  # doctest: +REMOTE_DATA
    >>> map2 = sunpy.map.Map(sunpy.data.sample.EIT_195_IMAGE)  # doctest: +REMOTE_DATA
    >>> mc = sunpy.map.Map([map1, map2], sequence=True)  # doctest: +REMOTE_DATA

or by providing a directory full of image files:

.. code-block:: python

    >>> mc = sunpy.map.Map('path/to/my/files/*.fits', sequence=True)   #  doctest: +SKIP

The earliest map in the MapSequence can be accessed by indexing the maps list:

.. code-block:: python

    >>> mc.maps[0]   # doctest: +SKIP

MapSequences can hold maps that have different shapes.
To test if all the maps in a `~sunpy.map.MapSequence` have the same shape:

.. code-block:: python

    >>> mc.all_maps_same_shape()  # doctest: +REMOTE_DATA
    True

It is often useful to return the image data in a `~sunpy.map.MapSequence` as a single three dimensional NumPy `~numpy.ndarray`:

.. code-block:: python

    >>> mc_array = mc.as_array()   # doctest: +REMOTE_DATA

Note that an array is returned only if all the maps have the same shape.

> If this is not true, a `ValueError` is returned.

Reviewer 2: ```suggestion
If this is not true, a `ValueError` is raised.
```
